### DISCUSSION TITLE
SPARQL construct - not finishing on larger data

### DISCUSSION THREAD

**Discussion by martina-parkanova** [linkedpipes/etl#987] [Q&A] [answered]
Hello, could you please help me with another problem with larger data? 
I have a file, originally json, changed to RDF with components JSON to JSON-LD and Files to RDF single graph. This file contains information about all current educational insitutions in ČR. In the next step, I used component SPARQL construct and tried to filter out only those institutions that are in "Pardubický kraj". I am not sure, if this is correct approach, if not, just tell me that I am totally wrong... As I want also RDF as my output file, but elements with their children as well, my SPARQL looks like this:

```sparql
PREFIX sp: <http://spinrdf.org/sp#>
PREFIX msmt: <http://msmt.cz/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

CONSTRUCT {
   ?parent ?p3 ?sub .
   ?parent ?p4 ?o4 .
   ?sub ?p ?o .
   ?o ?p2 ?o2 .
   ?o2 ?p5 ?o3 .
}
WHERE {
     ?sub ?p ?o .
     ?sub msmt:kraj ?kraj .
     FILTER(?kraj = 'Pardubický kraj') .
     OPTIONAL {?o ?p2 ?o2} .
     OPTIONAL {?o2 ?p5 ?o3} .
     ?parent ?p3 ?sub .
     ?parent ?p4 ?o4 .
}
```
[data.zip](https://github.com/linkedpipes/etl/files/15334283/data.zip)

Input data are attached...

**Comment by jakubklimek** [comment, marked as answer]
The query is quite generic like this, causing lots of joins and poor performance. It would be better to fit the predicates and classes present in the data.
Also, careful of the OPTIONALS as like this, you are doing a cartesian product as ?o ?p2 ?o2 is independent of ?o2 ?p5 ?o3.

This one without the parents works quite well:
```sparql
PREFIX sp: <http://spinrdf.org/sp#>
PREFIX msmt: <http://msmt.cz/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

CONSTRUCT {
   ?sub ?p ?o .
   ?o ?p2 ?o2 .
   ?o2 ?p5 ?o3 .
}
WHERE {
     ?sub ?p ?o .
     ?sub msmt:kraj ?kraj .
     FILTER(?kraj = 'Pardubický kraj') .
     OPTIONAL {?o ?p2 ?o2 OPTIONAL {?o2 ?p5 ?o3} } .
     
}
```

**Comment by martina-parkanova** [comment]
Thank you very much for your answer, it helped a lot. I am learning sparql here and wasn't aware about the optionals problem... Without the parents clause, it runs ok... with the parents, it is significantly slower... but finishes. I think I need the parents as well, but maybe I'll find a way how to live without them.
